tests/test_ledger.py

```diff
@@ -1,445 +1,147 @@
-        "Expecting one new line per currency and transaction"
-    )
-    assert sum(spit_txn["account"] == transitory_account) == id_currency_pairs, (
-        "Expecting one transaction on transitory account per id and currency"
-    )
-    assert all(spit_txn.groupby("id")["base_currency_amount"].sum() == 0), (
-        "Expecting split transactions to be balanced"
-    )
-    assert spit_txn.query("account == @transitory_account")["base_currency_amount"].sum() == 0, (
-        "Expecting transitory account to be balanced"
-    )
-
-
-def test_mirror_ledger(set_up_vat_and_account):
-    cashctrl = ExtendedCashCtrlLedger(transitory_account = 19999)
-
-    # Mirror with one single and one collective transaction
-    target = LEDGER_ENTRIES.query("id in [1, 2]")
-    cashctrl.mirror_ledger(target=target, delete=True)
-    expected = cashctrl.standardize_ledger(target)
-    mirrored = cashctrl.ledger()
-    assert txn_to_str(mirrored) == txn_to_str(expected)
-
-    # Mirror with duplicate transactions and delete=False
-    target = pd.concat(
-        [
-            LEDGER_ENTRIES.query("id == 1"),
-            LEDGER_ENTRIES.query("id == 1").assign(id=5),
-            LEDGER_ENTRIES.query("id == 2").assign(id=6),
-            LEDGER_ENTRIES.query("id == 2"),
+    def test_add_ledger_entry(self, ledger, ledger_id):
+        ledger.restore(accounts=self.ACCOUNTS, vat_codes=self.VAT_CODES)
+        target = self.LEDGER_ENTRIES.query("id == @ledger_id")
+        id = ledger.add_ledger_entry(target)
+        remote = ledger.ledger()
+        created = remote.loc[remote["id"] == str(id)]
+        expected = ledger.standardize_ledger(target)
+        assert_frame_equal(
+            created, expected, ignore_index=True, ignore_columns=["id"], check_exact=True
+        )
+
+    def test_add_ledger_with_non_existing_vat(self, ledger):
+        # Adding a ledger entry with non existing VAT code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["vat_code"].iat[0] = "Test_Non_Existent_VAT_code"
+        with pytest.raises(ValueError, match="No id found for tax code"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_account(self, ledger):
+        # Adding a ledger entry with non existing account should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["account"] = 33333
+        with pytest.raises(ValueError, match="No id found for account"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_currency(self, ledger):
+        # Adding a ledger entry with non existing currency code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["currency"] = "Non_Existent_Currency"
+        with pytest.raises(ValueError, match="No id found for currency"):
+            ledger.add_ledger_entry(target)
+
+    @pytest.mark.parametrize("id", [15, 16])
```

Comment:
This test might be easier to understand if the two ledger entries were hard coded right here (in csv format) .